ppp_hal/requesthandler.py

```diff
@@ -24,38 +24,128 @@ def connect_memcached():
     mc = memcache.Client(Config().memcached_servers)
     return mc
 
-def _query(query, fields):
-    params = {'q': query, 'wt': 'json', 'fl': fields}
-    streams = (requests.get(url, params=params, stream=True)
+class APIS:
+    PAPERS = 'search'
+    AUTHORS = 'ref/author'
+
+def _query(query, fields, api):
+    params = {'q': query, 'wt': 'json', 'fl': ','.join(fields)}
+    streams = (requests.get('%s/%s/' % (url, api), params=params, stream=True)
                for url in Config().apis)
     docs_lists = (s.json()['response']['docs'] for s in streams)
     return list(itertools.chain.from_iterable(docs_lists))
 
-def query(query, fields):
+def query(query, fields, api=APIS.PAPERS):
     mc = connect_memcached()
-    key = 'ppp-hal-%s' + hashlib.md5(pickle.dumps((query, fields))).hexdigest()
+    key = (query, fields, api)
+    key = 'ppp-hal-%s' + hashlib.md5(pickle.dumps(key)).hexdigest()
     r = mc.get(key)
     if not r:
-        r = _query(query, fields)
+        r = _query(query, fields, api)
         mc.set(key, r, time=Config().memcached_timeout)
     return r
 
+
+AUTHOR_FIELDS = ('url_s', 'email_s',
+        'firstName_s', 'lastName_s', 'fullName_s')
+
+def author_graph_from_docid(docid):
+    authors = query('docid:%s' % docid, AUTHOR_FIELDS, api=APIS.AUTHORS)
+    assert len(authors) == 1, authors
+    author = authors[0]
+    urls = author.get('url_s', [])
+    emails = author.get('email_s', [])
+    if isinstance(urls, str):
+        urls = [urls]
+    # Some researchers put their lab's URL instead of a personal URL.
+    # This heuristic seems to have no false positives (but a few
+    # false negatives)
+    urls = [x for x in urls if author['lastName_s'].lower() in x.lower()]
+
+    if isinstance(emails, str):
+        emails = [emails]
+    emails = ['mailto:%s' % x for x in emails]
+    uris = emails + urls
+    d = {'@context': 'http://schema.org',
+         '@type': 'Person',
+         'name': author['fullName_s'],
+         'givenName': author['firstName_s'],
+         'familyName': author['lastName_s']}
+    if uris:
+        d['@id'] = uris[0]
```

Comment:
It's not very nice. @id is the URI of the resource in the current graph. If you have no URIs for authors in HAL, just add no @id and use sameAs